File: FirstApp/Stock/ListTypeDgv.vb
Public Class ListTypeDgv

    Public Property ListItems As New List(Of String)

    Private Sub LoadItems()
        Dgv.Rows.Clear()
        For Each i In ListItems
            Dgv.Rows.Add()
            Dgv.Rows(Dgv.Rows.Count - 1).Cells(0).Value = Dgv.Rows.Count
            Dgv.Rows(Dgv.Rows.Count - 1).Cells(1).Value = i
        Next
        ItemTB.Text = ""
        ItemTB.Select()
    End Sub

    Private Sub AddBT_Click(sender As Object, e As EventArgs) Handles AddBT.Click
        If ItemTB.Text.Length <= 0 Then Exit Sub
        If ListItems.Contains(ItemTB.Text) Then MessageBox.Show("All Items Must Be Distinct.")
        ListItems.Add(ItemTB.Text)
        LoadItems()
    End Sub

    Private Sub EditBT_Click(sender As Object, e As EventArgs) Handles EditBT.Click
        Dim CurrentIndex As Integer = Dgv.CurrentRow.Index
        If CurrentIndex < 0 Then Exit Sub
        If ItemTB.Text.Length <= 0 Then Exit Sub
        ListItems(CurrentIndex) = ItemTB.Text
        LoadItems()
    End Sub

    Private Sub DeleteBT_Click(sender As Object, e As EventArgs) Handles DeleteBT.Click
        If Dgv.CurrentRow.Index < 0 Then Exit Sub
        ListItems.Remove(Dgv.CurrentRow.Cells(1).Value)
        LoadItems()
    End Sub

    Private Sub UpBT_Click(sender As Object, e As EventArgs) Handles UpBT.Click
        If Dgv.CurrentRow.Index <= 0 Then Exit Sub
        Dim newIndex As Integer = Dgv.CurrentRow.Index - 1
        Dim currentItem As String = Dgv.CurrentRow.Cells(1).Value
        DeleteBT_Click(sender, e)
        ListItems.Insert(newIndex, currentItem)
        LoadItems()
        Dgv.Rows(newIndex).Cells(0).Selected = True
    End Sub

    Private Sub DownBT_Click(sender As Object, e As EventArgs) Handles DownBT.Click
        If Dgv.CurrentRow.Index = Dgv.Rows.Count - 1 Then Exit Sub
        Dim newIndex As Integer = Dgv.CurrentRow.Index + 1
        Dim currentItem As String = Dgv.CurrentRow.Cells(1).Value
        DeleteBT_Click(sender, e)
        ListItems.Insert(newIndex, currentItem)
        LoadItems()
        Dgv.Rows(newIndex).Cells(0).Selected = True
    End Sub

    Private Sub Dgv_SelectionChanged(sender As Object, e As EventArgs) Handles Dgv.SelectionChanged
        If Dgv.CurrentRow.Index < 0 Then Exit Sub
        ItemTB.Text = Dgv.CurrentRow.Cells(1).Value
        ItemTB.Select()
    End Sub

    Public Event Submit_CLick()

    Private Sub SubmitBT_Click(sender As Object, e As EventArgs) Handles SubmitBT.Click
        RaiseEvent Submit_CLick()
    End Sub

    Public Event Cancel_CLick()
    Private Sub CancelBT_Click(sender As Object, e As EventArgs) Handles CancelBT.Click
        RaiseEvent Cancel_CLick()
    End Sub

    Private Sub ListTypeDgv_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        LoadItems()
    End Sub
End Class

File: FirstApp/Stock/UserControl/ListTypeDgv.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic.CompilerServices;

namespace FirstApp
{
    public partial class ListTypeDgv
    {

        public List<string> ListItems { get; set; } = new List<string>();

        public ListTypeDgv()
        {
            InitializeComponent();
            Dgv = _Dgv;
            _Dgv.Name = "Dgv";
        }

        private void LoadItems()
        {
            Dgv.Rows.Clear();
            foreach (var i in ListItems)
            {
                Dgv.Rows.Add();
                Dgv.Rows[Dgv.Rows.Count - 1].Cells[0].Value = Dgv.Rows.Count;
                Dgv.Rows[Dgv.Rows.Count - 1].Cells[1].Value = i;
            }
            ItemTB.Text = "";
            ItemTB.Select();
        }

        private void AddBT_Click(object sender, EventArgs e)
        {
            if (ItemTB.Text.Length <= 0)
                return;
            if (ListItems.Contains(ItemTB.Text))
                MessageBox.Show("All Items Must Be Distinct.");
            ListItems.Add(ItemTB.Text);
            LoadItems();
        }

        private void EditBT_Click(object sender, EventArgs e)
        {
            int CurrentIndex = Dgv.CurrentRow.Index;
            if (CurrentIndex < 0)
                return;
            if (ItemTB.Text.Length <= 0)
                return;
            ListItems[CurrentIndex] = ItemTB.Text;
            LoadItems();
        }

        private void DeleteBT_Click(object sender, EventArgs e)
        {
            if (Dgv.CurrentRow.Index < 0)
                return;
            ListItems.Remove(Conversions.ToString(Dgv.CurrentRow.Cells[1].Value));
            LoadItems();
        }

        private void UpBT_Click(object sender, EventArgs e)
        {
            if (Dgv.CurrentRow.Index <= 0)
                return;
            int newIndex = Dgv.CurrentRow.Index - 1;
            string currentItem = Conversions.ToString(Dgv.CurrentRow.Cells[1].Value);
            DeleteBT_Click(sender, e);
            ListItems.Insert(newIndex, currentItem);
            LoadItems();
            Dgv.Rows[newIndex].Cells[0].Selected = true;
        }

        private void DownBT_Click(object sender, EventArgs e)
        {
            if (Dgv.CurrentRow.Index == Dgv.Rows.Count - 1)
                return;
            int newIndex = Dgv.CurrentRow.Index + 1;
            string currentItem = Conversions.ToString(Dgv.CurrentRow.Cells[1].Value);
            DeleteBT_Click(sender, e);
            ListItems.Insert(newIndex, currentItem);
            LoadItems();
            Dgv.Rows[newIndex].Cells[0].Selected = true;
        }

        private void Dgv_SelectionChanged(object sender, EventArgs e)
        {
            if (Dgv.CurrentRow.Index < 0)
                return;
            ItemTB.Text = Conversions.ToString(Dgv.CurrentRow.Cells[1].Value);
            ItemTB.Select();
        }

        public event Submit_CLickEventHandler Submit_CLick;

        public delegate void Submit_CLickEventHandler();

        private void SubmitBT_Click(object sender, EventArgs e)
        {
            Submit_CLick?.Invoke();
        }

        public event Cancel_CLickEventHandler Cancel_CLick;

        public delegate void Cancel_CLickEventHandler();
        private void CancelBT_Click(object sender, EventArgs e)
        {
            Cancel_CLick?.Invoke();
        }

        private void ListTypeDgv_Load(object sender, EventArgs e)
        {
            LoadItems();
        }
    }
}